crates/bevy_pbr/src/wireframe.rs

```diff
@@ -41,158 +31,101 @@ impl Plugin for WireframePlugin {
         app.register_type::<Wireframe>()
             .register_type::<WireframeConfig>()
             .init_resource::<WireframeConfig>()
-            .add_plugins((
-                ExtractResourcePlugin::<WireframeConfig>::default(),
-                ExtractComponentPlugin::<Wireframe>::default(),
-            ));
-
-        if let Ok(render_app) = app.get_sub_app_mut(RenderApp) {
-            render_app
-                .add_render_command::<Opaque3d, DrawWireframes>()
-                .init_resource::<SpecializedMeshPipelines<WireframePipeline>>()
-                .add_systems(Render, queue_wireframes.in_set(RenderSet::Queue));
-        }
-    }
-
-    fn finish(&self, app: &mut bevy_app::App) {
-        if let Ok(render_app) = app.get_sub_app_mut(RenderApp) {
-            render_app.init_resource::<WireframePipeline>();
-        }
+            .add_plugins(MaterialPlugin::<WireframeMaterial>::default())
+            .add_systems(Update, apply_global)
+            .add_systems(Update, apply_material);
     }
 }
 
-/// Controls whether an entity should rendered in wireframe-mode if the [`WireframePlugin`] is enabled
-#[derive(Component, Debug, Clone, Default, ExtractComponent, Reflect)]
+/// Toggles wireframe rendering for any entity it is attached to.
+///
+/// This requires the [`WireframePlugin`] to be enabled.
+#[derive(Component, Debug, Clone, Default, Reflect)]
 #[reflect(Component, Default)]
 pub struct Wireframe;
 
+/// Configuration resource for [`WireframePlugin`].
 #[derive(Resource, Debug, Clone, Default, ExtractResource, Reflect)]
 #[reflect(Resource)]
 pub struct WireframeConfig {
-    /// Whether to show wireframes for all meshes. If `false`, only meshes with a [Wireframe] component will be rendered.
+    /// Whether to show wireframes for all meshes.
+    /// If `false`, only meshes with a [`Wireframe`] component will be rendered.
     pub global: bool,
 }
 
-#[derive(Resource, Clone)]
-pub struct WireframePipeline {
-    mesh_pipeline: MeshPipeline,
-    shader: Handle<Shader>,
-}
-impl FromWorld for WireframePipeline {
-    fn from_world(render_world: &mut World) -> Self {
-        WireframePipeline {
-            mesh_pipeline: render_world.resource::<MeshPipeline>().clone(),
-            shader: WIREFRAME_SHADER_HANDLE.typed(),
-        }
+/// Applies the wireframe material to any mesh with a [`Wireframe`] component.
+fn apply_material(
+    mut commands: Commands,
+    mut materials: ResMut<Assets<WireframeMaterial>>,
+    wireframes: Query<Entity, (With<Wireframe>, Without<Handle<WireframeMaterial>>)>,
+) {
```

Comment:
```suggestion
    wireframes: Query<Entity, (With<Wireframe>, Without<Handle<WireframeMaterial>>)>,
    mut removed_wireframes: RemovedComponents<WireFrame>,
) {
    for e in removed_wireframes.iter() {
        if let Some(mut commands) = commands.get_entity(e) {
            commands.remove::<Handle<Wireframe>>();
        }
    }
```

unless i missed it you also need to handle the case where the component gets removed.

Reply:
Done.